### PR TITLE
Add: run 43-layer DSpark DSA-CP prefill

### PR DESCRIPTION
**Pull request by MaxwellF1**
- Add a rank-local 43-layer prefill forward that dispatches SWA, CSA,
  and HCA layers from one loop, with the L3 host entry only allocating
  the communication windows and launching one child per rank.
- Replace the KV all-gather module with a context-parallel token-row
  all-gather that returns the TP group's rows in rank-major order on
  every rank, and rebuild the CSA/HCA/SWA CP attention entries around
  replicated layer boundaries with token-local compute.
- Keep o-projection weights persistently TP-sharded and reconstruct
  each layer into reusable full-weight scratch with explicit reuse
  ordering.
- Add a prefill MoE wave path with deferred dispatch waits, and
  synchronize and retire the CP all-gather and MoE retained-window
  epochs before their communication storage is reused.
- Add a single-layer numerical path that reuses the production
  attention and MoE stages for TP/CP group sizes 1, 2, and 4.

### CODE REVIEW

**Comment by MaxwellF1** [comment]
### Validation

- Numerical single-layer checks passed for SWA, CSA, and HCA,
  including their mutable caches and compression states.
- The MoE path passed two consecutive epochs using the same
  communication windows.
- A development-only exact check verified `o_proj` shard-to-full
  reconstruction for TP1, TP2, and TP4; that temporary utility is not
  included in this PR.
- The 43-layer native prefill completed on real devices for TP1, TP2,
  and TP4, including 8192 tokens on EP4/TP4.
- Python, Ruff, header, English-only, and public-document checks passed.

The 43-layer runner currently has no full-forward golden, so its result
is a liveness check. Numerical coverage is provided by the exact
communication checks and production-composed single-layer checks above.

This PR covers the 43-layer prefill backbone. Final normalization,
LM head, logits, and performance evaluation are outside its scope.

**Comment by MaxwellF1** [comment]
### A2/A3 follow-up

Fixed the `decode_layer.py` regression in `8cf31a21`.

The two-round MoE fixture had changed the default shape returned by the
shared `build_tensor_specs()` helper. The helper now preserves its
single-round contract for decode/prefill consumers, while the standalone
MoE runner explicitly requests two retained-window rounds.

- `decode_layer.py` completed numerical validation on two real devices.
- The two-round MoE fixture still completed exact numerical validation.
- Python, Ruff, and diff checks passed.

File: models/deepseek_v4_flash_dspark/prefill_fwd.py [hunk 1]
```diff
@@ -0,0 +1,2346 @@
...
+
+# Rank-sharded resident caches.
+RESIDENT_CACHE_NAMES = frozenset(CACHE_NAMES)
+
+# Caches returned to the following decode invocation.
+RESIDENT_CACHE_OUTPUT_NAMES = RESIDENT_CACHE_NAMES
+
+
+@pl.jit.inline(auto_scope=False)
+def gather_o_proj_full_weights(
```
Comment (zhangqi-chen) [inline models/deepseek_v4_flash_dspark/prefill_fwd.py]: move to an individual file if it is required by all attn, and include it with in all prefill_xxa.py

File: models/deepseek_v4_flash_dspark/prefill_fwd.py [hunk 2]
```diff
@@ -0,0 +1,2346 @@
...
+                    wo_b_local, [row, 0],
+                    [O_PROJ_WEIGHT_COPY_TILE, O_PROJ_FULL_ROWS],
+                    target_memory=pl.MemorySpace.Vec,
+                )
+                pl.store(tile, [row, 0], wo_b_full)
+        return pl.system.task_dummy(deps=[wo_a_copy_tid, wo_b_copy_tid])
+
+
+@pl.jit.inline(auto_scope=False)
+def _prefill_swa_attention(
```
Comment (zhangqi-chen) [inline models/deepseek_v4_flash_dspark/prefill_fwd.py]: delete these.

File: models/deepseek_v4_flash_dspark/prefill_fwd.py [hunk 3]
```diff
@@ -0,0 +1,2346 @@
...
+        arrived, data_arrived, routed_y_buf, combine_arrived,
+        stage_done, stage_token,
+        layer_id, my_rank, moe_epoch,
+    )
+    ffn_out[local_wave_base : local_wave_base + T, :] = ffn_wave
+    return barrier_tid
+
+
+@pl.jit.inline(auto_scope=False)
+def _prefill_moe(
```
Comment (zhangqi-chen) [inline models/deepseek_v4_flash_dspark/prefill_fwd.py]: prefill-moe and all its related should belongs to moe.py

File: models/deepseek_v4_flash_dspark/prefill_fwd.py [hunk 4]
```diff
@@ -0,0 +1,2346 @@
...
+    with pl.scope():
+        x_hc = prefill_moe_post(
+            ffn_out, attn_out, post_ffn, comb_ffn, x_hc,
+            gather_window, gather_signal, group_base, tp_rank,
+        )
+    return x_hc
+
+
+@pl.jit(auto_scope=False)
+def prefill_fwd(
```
Comment (zhangqi-chen) [inline models/deepseek_v4_flash_dspark/prefill_fwd.py:210]: prefill-fwd should only run 3 prefill_attn + moe, like models/deepseek_v4_flash_mtp/prefill_fwd.py

**Review by zhangqi-chen — CHANGES_REQUESTED** [review]

**Review by zhangqi-chen — APPROVED** [review]
